### PR TITLE
network/wifi: add getting statistics info from a wi-fi driver

### PR DESCRIPTION
**Pull request by pillip8282**
get wi-fi driver statistics information by wifi_manager_get_stats()

### CODE REVIEW

File: framework/src/wifi_manager/wifi_manager_state.c [hunk 1]
```diff
@@ -682,6 +674,34 @@ wifi_manager_result_e _handler_on_scanning_state(wifimgr_msg_s *msg)
 	return wret;
 }
 
+wifi_manager_result_e _handler_get_stats(wifimgr_msg_s *msg)
+{
+	wifi_manager_result_e wret = WIFI_MANAGER_SUCCESS;
+	trwifi_stats_s stats;
+	wifi_utils_result_e res = wifi_utils_get_stats(&stats);
+	if (res != WIFI_UTILS_SUCCESS) {
+		wret = WIFI_MANAGER_FAIL;
```
Comment (jeongchanKim) [inline framework/src/wifi_manager/wifi_manager_state.c:683]: In this case, we don't need to or should not set the information from stats in line from 685 to 700.
I think, we can just return the failure.

File: os/include/tinyara/net/if/wifi.h [hunk 2]
```diff
@@ -99,12 +99,29 @@ typedef struct trwifi_scan_list {
 } trwifi_scan_list_s;
 
 typedef struct {
-	uint32_t ip4_address;								   /**<	 ip4 address							   */
-	unsigned char mac_address[TRWIFI_MACADDR_LEN];		/**<  MAC address of wifi interface				*/
-	int rssi;											   /**<	 Receive Signal Strength Indication in dBm */
-	trwifi_status_e wifi_status;						/**<  @ref trwifi_status					 */
+	int rssi;						/**<	 Receive Signal Strength Indication in dBm */
+	trwifi_status_e wifi_status;	/**<  @ref trwifi_status					 */
+	struct timeval disconn_time; // optional: stored the last time when the connection is disconnected
```
Comment (abhishek-samsung) [inline os/include/tinyara/net/if/wifi.h:104]: Please change "stored" -->> "store"

File: os/include/tinyara/wifi/wifi_common.h [hunk 3]
```diff
@@ -186,13 +186,13 @@ struct wifi_utils_cb {
...
 /**
- * @brief wifi information (ip address, mac address)
+ * @brief wifi information
  */
 struct wifi_utils_info {
-	uint32_t ip4_address;                                 /**<  ip4 address                               */
-	unsigned char mac_address[WIFI_UTILS_MACADDR_LEN];    /**<  MAC address of wifi interface             */
 	int rssi;                                             /**<  Receive Signal Strength Indication in dBm */
 	wifi_utils_status_e wifi_status;                      /**<  @ref wifi_utils_status                    */
+	struct timeval disconn_time; // optional: stored the last time when the connection is disconnected
```
Comment (abhishek-samsung) [inline os/include/tinyara/wifi/wifi_common.h:194]: Please change "stored" -->> "store"

File: os/drivers/wireless/scsc/slsi_drv_netmgr.c [hunk 4]
```diff
@@ -53,26 +53,26 @@
...
-	sem_t scan_sem;
-	uint8_t scan_ssid[TRWIFI_SSID_LEN + 1];
-	trwifi_scan_list_s *result_list;
-};
+		struct _wifi_scan_filter_result_ {
+			int scan_flag;
+			sem_t scan_sem;
+			uint8_t scan_ssid[TRWIFI_SSID_LEN + 1];
+			trwifi_scan_list_s *result_list;
+		};
```
Comment (abhishek-samsung) [inline os/drivers/wireless/scsc/slsi_drv_netmgr.c:75]: Please check the indentation.

File: os/drivers/wireless/scsc/slsi_drv_netmgr.c [hunk 5]
```diff
@@ -599,7 +599,7 @@ trwifi_result_e slsidrv_connect_ap(struct netdev *dev, trwifi_ap_config_s *ap_co
 	} else {
 		result = TRWIFI_SUCCESS;
 		vdvdbg("Successfully joined the network: %s(%d)\n", ap_connect_config->ssid,
-			  ap_connect_config->ssid_length);
+			   ap_connect_config->ssid_length);
```
Comment (abhishek-samsung) [inline os/drivers/wireless/scsc/slsi_drv_netmgr.c:602]: Please check the indentation. There are extra spaces before the line.

File: os/drivers/wireless/virtual/vdev.c [hunk 6]
```diff
@@ -68,6 +69,7 @@ static struct trwifi_ops g_trwifi_drv_ops = {
 	vdev_start_softap,           /* start_softap */
 	vdev_stop_softap,            /* stop_softap */
 	vdev_set_autoconnect,        /* set_autoconnect */
+	vdev_get_stats,
```
Comment (deepaksrma) [inline os/drivers/wireless/virtual/vdev.c:72]: could you give a description for this as well?

File: framework/src/wifi_manager/wifi_manager_event.h [hunk 7]
```diff
@@ -30,6 +30,7 @@ enum _wifimgr_evt {
 	EVT_DISCONNECT_CMD,			// Command to Disconnect from a connected WiFi AP
 	EVT_SCAN_CMD,				// Command to perform WiFi Scanning over WLAN channels
 	EVT_GETINFO_CMD,			// Command to get WiFi Manager information
+	EVT_GETSTATS_CMD,           // Command to get WiFi driver stats
```
Comment (amogh-samsung) [inline framework/src/wifi_manager/wifi_manager_event.h:33]: Could you align these comment lines please?

File: os/drivers/wireless/scsc/slsi_drv_netmgr.c (hunk 5 above)
Comment (amogh-samsung) [reply os/drivers/wireless/scsc/slsi_drv_netmgr.c:602]: I think it's alright,
It's not a new line of code but a continued one, hence the extra 2 spaces.

File: os/include/tinyara/net/if/wifi.h [hunk 8]
```diff
@@ -306,6 +339,8 @@ typedef trwifi_result_e (*trwifi_set_autoconnect)(struct netdev *dev, uint8_t ch
  */
 typedef trwifi_result_e (*trwifi_drv_ioctl)(struct netdev *dev, int cmd, unsigned long arg);
 
+
```
Comment (amogh-samsung) [inline os/include/tinyara/net/if/wifi.h:342]: Do we need these extra lines here please?

File: os/drivers/wireless/virtual/vdev.c [hunk 9]
```diff
@@ -346,6 +347,29 @@ trwifi_result_e vdev_set_autoconnect(struct netdev *dev, uint8_t check)
 	return TRWIFI_SUCCESS;
 }
 
+trwifi_result_e vdev_get_stats(struct netdev *dev, trwifi_stats_s *stats)
```
Comment (jeongchanKim) [inline os/drivers/wireless/virtual/vdev.c:350]: `dev` is not used anywhere. Is it needed?

File: os/include/tinyara/net/if/wifi.h [hunk 10]
```diff
@@ -99,12 +99,29 @@ typedef struct trwifi_scan_list {
 } trwifi_scan_list_s;
 
 typedef struct {
-	uint32_t ip4_address;								   /**<	 ip4 address							   */
-	unsigned char mac_address[TRWIFI_MACADDR_LEN];		/**<  MAC address of wifi interface				*/
-	int rssi;											   /**<	 Receive Signal Strength Indication in dBm */
-	trwifi_status_e wifi_status;						/**<  @ref trwifi_status					 */
+	int rssi;						/**<	 Receive Signal Strength Indication in dBm */
```
Comment (dineshdas-samsung) [inline os/include/tinyara/net/if/wifi.h:102]: Did you mean  Received Signal Strength Indicator (RSSI) ?

File: os/include/tinyara/wifi/wifi_common.h [hunk 11]
```diff
@@ -186,13 +186,13 @@ struct wifi_utils_cb {
 typedef struct wifi_utils_cb wifi_utils_cb_s;
 
 /**
- * @brief wifi information (ip address, mac address)
+ * @brief wifi information
  */
 struct wifi_utils_info {
-	uint32_t ip4_address;                                 /**<  ip4 address                               */
-	unsigned char mac_address[WIFI_UTILS_MACADDR_LEN];    /**<  MAC address of wifi interface             */
 	int rssi;                                             /**<  Receive Signal Strength Indication in dBm */
```
Comment (dineshdas-samsung) [inline os/include/tinyara/wifi/wifi_common.h:192]: Please check once !

File: framework/src/wifi_manager/wifi_manager_event.h [hunk 12]
```diff
@@ -30,10 +30,11 @@ enum _wifimgr_evt {
 	EVT_DISCONNECT_CMD,			// Command to Disconnect from a connected WiFi AP
 	EVT_SCAN_CMD,				// Command to perform WiFi Scanning over WLAN channels
 	EVT_GETINFO_CMD,			// Command to get WiFi Manager information
+	EVT_GETSTATS_CMD,			// Command to get WiFi driver stats
 	EVT_RECONNECT_CMD,			// Command to reconnect to WiFi AP. If we use EVT_CONNECT_CMD, then we can't distinguish it's from application request
-	EVT_STA_CONNECTED,		// Event that STA is connected
+	EVT_STA_CONNECTED,			// Event that STA is connected
 	EVT_STA_CONNECT_FAILED, // Event that STA connect failed
-	EVT_STA_DISCONNECTED,   // Event that external STA disconnected from WiFi AP
+	EVT_STA_DISCONNECTED,	// Event that external STA disconnected from WiFi AP
```
Comment (dineshdas-samsung) [inline framework/src/wifi_manager/wifi_manager_event.h:37]: Can we please align the comments?

File: apps/examples/wifi_manager_sample/wm_test.c [hunk 13]
```diff
@@ -314,41 +316,41 @@ test_func func_table[WM_TEST_MAX] = {
...
-	"stats",
-	"info",
-	"auto",
+	"start",                                /* WM_TEST_START    */
+	"stop",								    /* WM_TEST_STOP     */
+	"softap",							    /* WM_TEST_SOFTAP  */
+	"sta",								    /* WM_TEST_STA      */
+	"join",								    /* WM_TEST_JOIN    */
+	"leave",							    /* WM_TEST_LEAVE    */
+	"cancel",							    /* WM_TEST_CANCEL   */
```
Comment (dineshdas-samsung) [inline apps/examples/wifi_manager_sample/wm_test.c:343]: Can we please confirm whether comments are aligned?

Comment (pillip8282) [reply apps/examples/wifi_manager_sample/wm_test.c:343]: aligning the comments is different to each editor. So I don't want fix it

File: framework/src/wifi_manager/wifi_manager_event.h (hunk 12 above)
Comment (pillip8282) [reply framework/src/wifi_manager/wifi_manager_event.h:37]: aligning the comments is different to each editor. So I don't want fix it

**Review by sangwon03 — APPROVED** [review]
LGTM

**Pull request merged**